daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/transaction/NormalizeRollbacks.scala

```diff
@@ -0,0 +1,188 @@
+          // no rule
+          List(Norm.Roll2(h, forceNorms(m), t))
+      }
+    }
+
+    def pushIntoRoll(a1: Norm.Act, xs2: List[Node], t: Norm.Roll): Norm.Roll = {
+      t match {
+        case Norm.Roll1(a3) => Norm.Roll2(a1, xs2, a3)
+        case Norm.Roll2(a3, xs4, a5) => Norm.Roll2(a1, xs2 ++ List(a3.node) ++ xs4, a5)
+      }
+    }
+
+    // Here we traverse the original transaction structure.
+    // During the transformation, an original `Node` is mapped into a List[Norm]
+
+    // The List is necessary because the rules can:
+    // (1) drop nodes; (2) combine nodes (3) lift nodes from a lower level to a higher level.
+
+    // TODO: https://github.com/digital-asset/daml/issues/8020
+    // make this traversal code stack-safe
+
+    txOriginal match {
+      case GenTransaction(nodesOriginal, rootsOriginal) =>
+        def traverseNids(nids: List[Nid]): List[Norm] = nids.flatMap(traverseNid)
+        def traverseNid(nid: Nid): List[Norm] = traverseNode(nodesOriginal(nid))
+
+        def traverseNode(node: Node): List[Norm] = {
+          node match {
+
+            case NodeRollback(children) =>
+              makeRoll(traverseNids(children.toList))
+
+            case ex: NodeExercises[_, _] =>
+              val norms = traverseNids(ex.children.toList)
+              val children = pushNodes(forceNorms(norms))
+              List(Norm.Act(ex.copy(children = ImmArray(children))))
+
+            case act: LeafOnlyActionNode[_] =>
+              List(Norm.Act(act))
+          }
+        }
+        val norms = traverseNids(rootsOriginal.toList)
+        val roots = pushNodes(forceNorms(norms))
+        GenTransaction(nodes, ImmArray(roots))
+    }
+  }
+
+  // Types which ensure we can only represent the properly normalized cases.
+  object Canonical {
+
+    // A properly normalized Tx/node
+    sealed trait Norm
+    object Norm {
+
+      // A non-rollback tx/node
+      final case class Act(node: ActNode) extends Norm
+
+      // A *normalized* rollback tx/node. 2 cases:
+      // - rollback containing a single non-rollback tx/node.
+      // - rollback of 2 or more tx/nodes, such that first and last are not rollbacks.
+      sealed trait Roll extends Norm
+      final case class Roll1(act: Act) extends Roll
+      final case class Roll2(head: Act, middle: List[Node], tail: Act) extends Roll
+    }
+
+    // Case analysis on a list of Norms, distinuishing: Empty, Single and Multi forms
+    // The Multi form separes the head and tail element for the middle-list.
+    sealed trait Case
+    object Case {
+      final case object Empty extends Case
+      final case class Single(n: Norm) extends Case
+      final case class Multi(h: Norm, m: List[Norm], t: Norm) extends Case
+    }
+
+    def caseNorms(xs: List[Norm]): Case = {
+      xs.length match {
+        case 0 => Case.Empty
+        case 1 => Case.Single(xs(0))
+        case n => Case.Multi(xs(0), xs.slice(1, n - 1), xs(n - 1))
```

Comment:
Maybe worth using a `Vector` or an `ImmArray` instead of slicing & calling length on a single-linked list? @remyhaemmerle-da probably has a better idea what makes sense here.

Reply:
calling length on List is probably a good idea. 
Why not using standard List pattern matching ?

Reply:
switched to Vector to address this concern